watcher/local/watcher.go

```diff
@@ -95,28 +106,156 @@ func (w *Watcher) startWatching(ctx context.Context, parent channel.ID, signedSt
-	thisCh *ch) {
+func (w *Watcher) handleEventsFromChain(eventsFromChainSub channel.AdjudicatorSubscription,
+	eventsToClientPubSub adjudicatorPub, thisCh *ch) {
+	parent := thisCh
+	if thisCh.parent != channel.Zero {
+		parent, _ = w.registry.retrieve(thisCh.parent)
+	}
+
 	for e := eventsFromChainSub.Next(); e != nil; e = eventsFromChainSub.Next() {
 		switch e.(type) {
 		case *channel.RegisteredEvent:
-			log := log.WithFields(log.Fields{"ID": e.ID(), "Version": e.Version()})
-			log.Debug("Received registered event from chain")
-			eventsToClientPubSub.Publish(e)
+			parent.subChsAccess.Lock()
+			func() {
+				defer parent.subChsAccess.Unlock()
+
+				log := log.WithFields(log.Fields{"ID": e.ID(), "Version": e.Version()})
+				log.Debug("Received registered event from chain")
+
+				eventsToClientPubSub.publish(e)
+
+				latestTx := thisCh.retreiveLatestTx()
+				log.Debugf("Latest version is (%d)", latestTx.Version)
+
+				if e.Version() < latestTx.Version {
+					if e.Version() < thisCh.registeredVersion {
+						log.Debugf("Latest version (%d) already registered ", thisCh.registeredVersion)
+						return
+					}
+					log.Debugf("Registering latest version (%d)", latestTx.Version)
+					err := registerDispute(w.registry, w.rs, parent)
+					if err != nil {
+						log.Error("Error registering dispute")
+						// TODO: Should the subscription be closed ?
+						return
+					}
+					log.Debug("Registered successfully")
+				}
+			}()
 		default:
 		}
 	}
 }
+
+// registerDispute collects the latest transaction for the parent channel and
+// each of its children. It then registers dispute for the channel tree.
+//
+// This function assumes the callers has locked the parent channel.
+func registerDispute(r *registry, registerer channel.Registerer, parentCh *ch) error {
+	parentTx := parentCh.retreiveLatestTx()
+	subStates := retreiveLatestSubStates(r, parentTx)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	err := registerer.Register(ctx, makeAdjudicatorReq(parentCh.params, parentTx), subStates)
+	if err != nil {
+		return err
+	}
+
+	parentCh.registeredVersion = parentTx.Version
+	for i := range subStates {
+		subCh, _ := r.retrieve(parentTx.Allocation.Locked[i].ID)
+		subCh.registeredVersion = subStates[i].State.Version
+	}
+	return nil
+}
+
+func retreiveLatestSubStates(r *registry, parentTx channel.Transaction) []channel.SignedState {
+	subStates := make([]channel.SignedState, len(parentTx.Allocation.Locked))
+	for i := range parentTx.Allocation.Locked {
+		// Can be done concurrently.
+		subCh, _ := r.retrieve(parentTx.Allocation.Locked[i].ID)
+		subChTx := subCh.retreiveLatestTx()
+		subStates[i] = channel.SignedState{
+			Params: subCh.params,
+			State:  subChTx.State,
+			Sigs:   subChTx.Sigs,
+		}
+	}
+	return subStates
+}
+
+func makeAdjudicatorReq(params *channel.Params, tx channel.Transaction) channel.AdjudicatorReq {
+	return channel.AdjudicatorReq{
```

Comment:
Just a comment: I added an entry for adding `NewAdjudicatorReq` in https://github.com/hyperledger-labs/go-perun/issues/179.

Reply:
I think, in that case, `NewAdjudicatorReq` should be implemented as a function that takes `params` and `Transaction` as parameter and constructs an `AdjudicatorReq` rather than being a method over channel type.

Because,
1. In case of client, it might never need this method.
2. In case of watcher, the watcher will not have access to channel instance, rather only to `params` and `Transaction`.

Reply:
Yes i agree, just wanted to let you know that we maybe introduce that function in the future.